include/swift/Reflection/ReflectionContext.h

```diff
@@ -244,11 +215,13 @@ class ReflectionContext
         if (strncmp(S->sectname, Name.data(), strlen(Name.data())) != 0)
           continue;
         auto RemoteSecStart = S->addr + Slide;
-        auto SectBufData = reinterpret_cast<const char *>(SectBuf.get());
-        auto LocalSectStart =
-            reinterpret_cast<const char *>(SectBufData + RemoteSecStart - RangeStart);
-        
-        auto StartRef = RemoteRef<void>(RemoteSecStart, LocalSectStart);
+        auto LocalSectBuf =
+            this->getReader().readBytes(RemoteAddress(RemoteSecStart), S->size);
+        if (!LocalSectBuf)
+          return {nullptr, 0};
```

Comment:
I was curious about this, whether a problem with one section could cause issues that previous didn't exist. It doesn't look like it, but I do see something that looks like it could be a problem in the surrounding code.

This lambda is used to read the following sections:

```
__swift5_fieldmd
__swift5_assocty
__swift5_builtin
__swift5_capture
__swift5_typeref
__swift5_reflstr
```

If any of these sections are missing or can't be read, then `readMachOSections` returns early:

```cpp
if (FieldMdSec.first == nullptr &&
    AssocTySec.first == nullptr &&
    BuiltinTySec.first == nullptr &&
    CaptureSec.first == nullptr &&
    TypeRefMdSec.first == nullptr &&
    ReflStrMdSec.first == nullptr)
  return false;

ReflectionInfo info = {
    {FieldMdSec.first, FieldMdSec.second},
    {AssocTySec.first, AssocTySec.second},
    {BuiltinTySec.first, BuiltinTySec.second},
    {CaptureSec.first, CaptureSec.second},
    {TypeRefMdSec.first, TypeRefMdSec.second},
    {ReflStrMdSec.first, ReflStrMdSec.second}};

this->addReflectionInfo(info);
```

### Question

I looked at my small swift program, and it has only 3 of these sections, it does not have:
* `__swift5_assocty`
* `__swift5_builtin`
* `__swift5_reflstr`

I am not especially familiar with these sections, but I can imagine there being some images that don't use associated types.

In practice, can these sections truly be absent, and, what does this break when one of these sections is missing?

Reply:
There's a comment in a test that says `__swift5_reflstr` is optional, possibly stripped:

https://github.com/apple/swift/blob/2b435f9839e10ea5901b1fa560f7bb87650bfbaa/stdlib/private/SwiftReflectionTest/SwiftReflectionTest.swift#L96

Reply:
That logic skips the image only if all sections are missing. If any are present, then it gets added.

Reply:
oh facepalm haha

Reply:
can I delete these all so nobody sees how poor my reading comprehension is?

Reply:
Sure, but we've all been there, no worries.

Reply:
yup! I was kidding about deleting :D